lib/auth/version.go

```diff
@@ -20,55 +20,101 @@ package auth
 
 import (
 	"context"
+	"log/slog"
 	"os"
 
 	"github.com/coreos/go-semver/semver"
 	"github.com/gravitational/trace"
+
+	authinfov1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/authinfo/v1"
+	"github.com/gravitational/teleport/api/types/authinfo"
+	"github.com/gravitational/teleport/lib/services"
 )
 
 const (
 	// skipVersionUpgradeCheckEnv is environment variable key for disabling the check
 	// major version upgrade check.
 	skipVersionUpgradeCheckEnv = "TELEPORT_UNSTABLE_SKIP_VERSION_UPGRADE_CHECK"
+	// versionUpgradeCheckMaxWriteRetry is the number of retries for conditional updates of the AuthInfo resource.
+	versionUpgradeCheckMaxWriteRetry = 5
 )
 
 // validateAndUpdateTeleportVersion validates that the major version persistent in the backend
 // meets our upgrade compatibility guide.
 func validateAndUpdateTeleportVersion(
 	ctx context.Context,
-	storage VersionStorage,
-	currentVersion *semver.Version,
-) error {
+	procStorage VersionStorage,
+	backendStorage services.AuthInfoService,
+	currentVersion semver.Version,
+) (err error) {
 	if skip := os.Getenv(skipVersionUpgradeCheckEnv); skip != "" {
 		return nil
 	}
```

Comment:
Should we forcibly update the value in the backend if the envvar is set? There's no good way to "heal" the value if there's some problem, otherwise.